src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+pub trait Message {
+    fn get_size(&self) -> i32;
+}
+
+pub trait FromByte: Default {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors>;
+
+    fn decode_new(buf: &mut Buf) -> Result<Self, DecodeErrors> {
+        let mut temp = Self::default();
+        match temp.decode(buf) {
+            Ok(_) => Ok(temp),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+pub trait DecodableMessage: Message + FromByte {}
+
+pub trait Request: DecodableMessage {
+    fn get_header(&self) -> &HeaderRequest;
+}
+
+// Skip for now
+pub trait Response: DecodableMessage {}
+
+impl FromByte for i8 {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors> {
+        *self = buf.get_i8();
```

Comment:
Note that if `decode` returned `Self`, the implementation of this method would just be 
```rust
fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
    buf.get_i8()
}
```

Which seems more idiomatic to me...